### PR TITLE
Aggregated discoveries

### PR DESCRIPTION
**Pull request by Felix6464**

### CODE REVIEW

File: pht_federated/aggregator/api/discoveries/statistics.py [hunk 1]
```diff
@@ -0,0 +1,225 @@
...
+from pandas.api.types import is_numeric_dtype, is_bool_dtype
+import plotly.express as px
+import plotly.io
+from plotly.graph_objects import Figure
+import json
+from pht_federated.aggregator.api.schemas.discovery import DiscoverySummary, DiscoveryFigure
+from pht_federated.aggregator.api.schemas.dataset_statistics import DatasetStatistics
+import plotly.graph_objects as go
+import math
+def get_discovery_statistics(dataframe: pd.DataFrame) -> Optional[DatasetStatistics]:
```
Comment (migraf) [inline pht_federated/aggregator/api/discoveries/statistics.py:12]: space between imports

File: pht_federated/aggregator/api/discoveries/statistics.py [hunk 2]
```diff
@@ -0,0 +1,225 @@
...
+    """
+    Computes statistical information of a dataset
+    :param dataframe: Dataset as dataframe object
+    :return: Dataset statistics
+    """
+    if not (isinstance(dataframe, pd.DataFrame)):
+        raise TypeError
```
Comment (migraf) [inline pht_federated/aggregator/api/discoveries/statistics.py:19]: Better error message

File: pht_federated/aggregator/api/endpoints/discovery.py [hunk 3]
```diff
@@ -0,0 +1,433 @@
...
+        discovery_feature_count /= len(response)
+
+        discovery_summary_schema = {
+            "proposal_id": proposal_id,
+            "item_count": discovery_item_count,
+            "feature_count": discovery_feature_count,
+            "data_information": aggregated_feature_lst
+        }
+
+        discovery_summary = DiscoverySummary(**discovery_summary_schema)
```
Comment (migraf) [inline pht_federated/aggregator/api/endpoints/discovery.py:224]: split into subfunctions and move to different file. function should have ~10 lines

File: pht_federated/aggregator/api/endpoints/discovery.py [hunk 4]
```diff
@@ -0,0 +1,433 @@
...
+
+        #print("DISCOVERY SUMMARY : {}".format(discovery_summary))
+
+        return discovery_summary
+
+@router.get("/{proposal_id}/discovery_feature", response_model=DiscoverySummary)
+def get_discovery_single(proposal_id: int,  feature_name: str, db: Session = Depends(dependencies.get_db)):
```
Comment (migraf) [inline pht_federated/aggregator/api/endpoints/discovery.py:231]: feature name can be an optional parameter of the discovery endpoint

File: pht_federated/aggregator/api/endpoints/discovery.py [hunk 5]
```diff
@@ -0,0 +1,433 @@
...
+                "value": discovery_equal_value
+            }
+            discovery_summary_schema = {
+                "proposal_id": proposal_id,
+                "item_count": discovery_item_count,
+                "feature_count": discovery_feature_count,
+                "data_information": [discovery_summary_json_equal]
+            }
+
+        discovery_summary = DiscoverySummary(**discovery_summary_schema)
```
Comment (migraf) [inline pht_federated/aggregator/api/endpoints/discovery.py:406]: see above -> outsource, split into functions, generalize

File: pht_federated/aggregator/api/endpoints/discovery.py [hunk 6]
```diff
@@ -0,0 +1,433 @@
...
+def post_discovery_statistics(proposal_id: int, create_msg: StatisticsCreate, db: Session = Depends(dependencies.get_db)) -> DatasetStatistics:
+    dataset = json.loads(create_msg.json())
+    discovery_statistics_schema = {
+        "proposal_id": proposal_id,
+        "n_items": int(dataset['n_items']),
+        "n_features": int(dataset['n_features']),
+        "column_information": dataset['column_information']
+    }
+    discovery_statistics = DiscoveryStatistics(**discovery_statistics_schema)
+    discovery_statistics = datasets.create(db, obj_in=discovery_statistics)
```
Comment (migraf) [inline pht_federated/aggregator/api/endpoints/discovery.py:429]: seems like unnecessary parsing, define input arguments and types directly/correctly.
why the int conversion?
dataset/s seems like the wrong name

datasets.create() should take StatisticsCreate as input

File: pht_federated/aggregator/api/endpoints/discovery.py [hunk 7]
```diff
@@ -0,0 +1,433 @@
...
+    if not discovery_statistics:
+        raise HTTPException(status_code=404, detail=f"DatasetStatistics of proposal with id '{proposal_id}' could not be created.")
```
Comment (migraf) [inline pht_federated/aggregator/api/endpoints/discovery.py:431]: 404 error means not found, either 400 or 500

File: pht_federated/aggregator/api/models/dataset_statistics.py [hunk 8]
```diff
@@ -0,0 +1,15 @@
+from pht_federated.aggregator.db.base_class import Base
+
+from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, JSON
+
+
+
+class DatasetStatistics(Base):
+    __tablename__ = "dataset_statistics"
+    id = Column(Integer, primary_key=True, index=True)
+    proposal_id = Column(Integer, default=0)
```
Comment (migraf) [inline pht_federated/aggregator/api/models/dataset_statistics.py:10]: this should be a foreign key to an additional `proposal` table

File: pht_federated/aggregator/api/models/discovery.py [hunk 9]
```diff
@@ -0,0 +1,15 @@
+from pht_federated.aggregator.db.base_class import Base
+
+from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, JSON
+
+
+
+class DiscoverySummary(Base):
+    __tablename__ = "discovery_summary"
+    id = Column(Integer, primary_key=True, index=True)
+    proposal_id = Column(Integer, default=0)
```
Comment (migraf) [inline pht_federated/aggregator/api/models/discovery.py:10]: foreign key

File: pht_federated/aggregator/api/models/discovery.py [hunk 10]
```diff
@@ -0,0 +1,15 @@
+from pht_federated.aggregator.db.base_class import Base
+
+from sqlalchemy import Column, Integer, String, DateTime, ForeignKey, JSON
+
+
+
+class DiscoverySummary(Base):
```
Comment (migraf) [inline pht_federated/aggregator/api/models/discovery.py:7]: this seems to be exactly the same as DatasetStatistics? Are booth needed?

File: pht_federated/aggregator/api/schemas/dataset_statistics.py [hunk 11]
```diff
@@ -0,0 +1,73 @@
...
+class DatasetUniqueColumn(BaseModel):
+    type: Literal['unique']
+    title: Optional[str]
+    number_of_duplicates: Optional[int]
+
+
+class DatasetEqualColumn(BaseModel):
+    type: Literal['equal']
+    title: Optional[str]
+    value: Optional[str]
```
Comment (migraf) [inline pht_federated/aggregator/api/schemas/dataset_statistics.py:17]: what does this do?

File: pht_federated/aggregator/api/schemas/dataset_statistics.py [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+    max: Optional[float]
+
+class DiscoveryUnstructuredColumn(BaseModel):
+    type: Literal['unstructured']
+    title: Optional[str]
+    not_na_elements: Optional[int]
+    number_targets: Optional[int]
+    target_counts: Optional[Dict[str, int]]
+    most_frequent_target: Optional[Union[int, str]]
+    frequency: Optional[int]
```
Comment (migraf) [inline pht_federated/aggregator/api/schemas/dataset_statistics.py:46]: Unstructured column? Datasets can be unstructured, columns only exist in structured data

File: pht_federated/aggregator/api/schemas/discovery.py [hunk 13]
```diff
@@ -0,0 +1,72 @@
...
+    class Config:
+        orm_mode = True
+
+
+class SummaryCreate(DiscoverySummary):
+    pass
+
+
+class SummaryUpdate(DiscoverySummary):
+    pass
```
Comment (migraf) [inline pht_federated/aggregator/api/schemas/discovery.py:70]: Duplicate of discovery statistics -> generalize

File: pht_federated/aggregator/api/schemas/figures.py [hunk 14]
```diff
@@ -0,0 +1,26 @@
...
+    title: Optional[str]
+    type: Optional[str]
+    figure: Optional[FigureData]
+
+class FigureCreate(DiscoveryFigure):
+    pass
+
+
+class FigureUpdate(DiscoveryFigure):
+    pass
```
Comment (migraf) [inline pht_federated/aggregator/api/schemas/figures.py:19]: Are these ever used?

File: pht_federated/aggregator/db/setup_db.py [hunk 15]
```diff
@@ -0,0 +1,44 @@
...
+        for _ in range(3):
+            dt = discovery.DataSet(
+                proposal_id=str(uuid.uuid4()),
+
+            )
+            dts.append(dt)
+        session.add_all(dts)
+        session.commit()
+
+    session.close()
```
Comment (migraf) [inline pht_federated/aggregator/db/setup_db.py:39]: is this needed

File: pht_federated/aggregator/storage/db/engine.py [hunk 16]
```diff
@@ -2,4 +2,6 @@
 
 ENGINE_URL = 'postgresql+psycopg2://postgres:postgres@localhost/postgres'
 
+ENGINE_URL = "postgresql+psycopg2://admin:admin@localhost:5442/pht_station"
+
 engine = create_engine(ENGINE_URL)
```
Comment (migraf) [inline pht_federated/aggregator/storage/db/engine.py:7]: seems like duplicate of `db/session.py`

File: pht_federated/client/storage/fs.py [hunk 17]
```diff
@@ -7,7 +7,7 @@
 
 class MinioBuckets(enum.Enum):
     MODELS = 'models'
-    DATASETS = 'datasets'
+    DATASETS = 'discoveries'
```
Comment (migraf) [inline pht_federated/client/storage/fs.py:10]: datasets = discoveries? :D why is this needed?

**Review by migraf — CHANGES_REQUESTED** [review]

File: pht_federated/client/storage/fs.py (hunk 17 above)
Comment (Felix6464) [reply pht_federated/client/storage/fs.py:10]: hab ich nie etwas geschrieben in dem file soweit ich weiß

Comment (Felix6464) [reply pht_federated/client/storage/fs.py:10]: steht bei mir auch dass du das als letzte Person verändert hast :D

File: pht_federated/aggregator/api/crud/crud_dataset_statistics.py [hunk 18]
```diff
@@ -0,0 +1,22 @@
...
+class CRUDDatasetStatistics(CRUDBase[DatasetStatistics, StatisticsCreate, StatisticsUpdate]):
+
+
+    def get_all_by_proposal_id(self, proposal_id: int, db: Session = Depends(dependencies.get_db)) -> List[DatasetStatistics]:
+        dataset = db.query(DatasetStatistics).filter(DatasetStatistics.proposal_id == proposal_id).all()
+        return dataset
+
+    def delete_by_proposal_id(self, proposal_id: int, db: Session = Depends(dependencies.get_db)) -> DatasetStatistics:
+        dataset_del = db.query(DatasetStatistics).filter(DatasetStatistics.proposal_id == proposal_id).delete()
+        return dataset_del
```
Comment (migraf) [inline pht_federated/aggregator/api/crud/crud_dataset_statistics.py:19]: return type correct? this query should match multiple items

File: pht_federated/aggregator/api/discoveries/statistics.py [hunk 19]
```diff
@@ -0,0 +1,226 @@
...
+
+def get_column_information(dataframe: pd.DataFrame, description: pd.DataFrame) -> dict:
+    """
+    Extract information out of dataframe and summarize it in a dictionary
+    :param dataframe: Dataframe to summarize
+    :param description: Dataframe description as dataframe object
+    :return: Dictionary with column information
+    """
+    columns_inf = []
+    columns = dataframe.columns.values.tolist()
```
Comment (migraf) [inline pht_federated/aggregator/api/discoveries/statistics.py:46]: iterate over column names directly `for col in dataframe:`

File: pht_federated/aggregator/api/discoveries/statistics.py [hunk 20]
```diff
@@ -0,0 +1,226 @@
...
+    :return: array with column information, key and json to save chart in cache
+    """
+    count = columns_inf[i]['not_na_elements']
+    unique = description[title]["unique"]
+    top = description[title]["top"]
+    freq = description[title]["freq"]
+
+
+    # if every entry has a unique value (or at most 50 values are given multiple times)
+    if count - 50 < unique <= count:
```
Comment (migraf) [inline pht_federated/aggregator/api/discoveries/statistics.py:104]: 50? should probably be percentage based

File: pht_federated/aggregator/api/discoveries/statistics.py [hunk 21]
```diff
@@ -0,0 +1,226 @@
...
+def create_figure(fig: Figure) -> DiscoveryFigure:
+    """
+    Create DataSetFigure-Object of a plotly figure
+    :param fig: Plotly figure
+    :return: DataSetFigure-Object with JSON-representation of plotly figure
+    """
+    fig_json = plotly.io.to_json(fig)
+    obj = json.loads(fig_json)
+    figure = DiscoveryFigure(fig_data=obj)
+    #print("Figure  data : {}".format(figure.fig_data))
```
Comment (migraf) [inline pht_federated/aggregator/api/discoveries/statistics.py:138]: use `DiscoveryFigure.parse_raw()` instead of manual to_json -> dict -> Figure

File: pht_federated/aggregator/api/discoveries/statistics.py [hunk 22]
```diff
@@ -0,0 +1,226 @@
...
+    return figure
+
+def plot_figure_json(json_data: dict):
+    fig_plotly = plotly.io.from_json(json.dumps(json_data))
+    fig_plotly.show()
```
Comment (migraf) [inline pht_federated/aggregator/api/discoveries/statistics.py:143]: unnecessary conversion `plotly.io.show(dict)`

File: pht_federated/aggregator/api/discoveries/utility_functions.py [hunk 23]
```diff
@@ -0,0 +1,158 @@
+from pht_federated.aggregator.api.discoveries.statistics import *
+from collections import Counter
+
+
+def aggregate_numeric_column(feature_lst: list, feature: dict):
```
Comment (migraf) [inline pht_federated/aggregator/api/discoveries/utility_functions.py:5]: function + variable names should be plural

File: pht_federated/aggregator/api/discoveries/utility_functions.py [hunk 24]
```diff
@@ -0,0 +1,158 @@
...
+        'type': "numeric",
+        'figure': obj
+    }
+
+    discovery_figure = DiscoveryFigure(**figure_schema)
+    discovery_summary_json['figure_data'] = discovery_figure
+
+    return discovery_summary_json
+
+def aggregate_categorical_column(feature_lst: list, feature: dict, num_discoveries: int):
```
Comment (migraf) [inline pht_federated/aggregator/api/discoveries/utility_functions.py:54]: plural

File: pht_federated/aggregator/api/discoveries/statistics.py [hunk 25]
```diff
@@ -0,0 +1,226 @@
...
+            columns_inf[i]['number_categories'] = unique
+            columns_inf[i]['most_frequent_element'] = top
+            columns_inf[i]['frequency'] = freq
+            columns_inf[i]['value_counts'] = value_counts
+
+        columns_inf[i]['type'] = column_type
+
+    return columns_inf
+
+def create_figure(fig: Figure) -> DiscoveryFigure:
```
Comment (migraf) [inline pht_federated/aggregator/api/discoveries/statistics.py:129]: move plot functions into separate file `plots.py`

File: pht_federated/aggregator/api/discoveries/utility_functions.py [hunk 26]
```diff
@@ -0,0 +1,158 @@
...
+    else:
+        raise Exception(f"Not all discoveries share an equal column considering the feature {discovery_title} ")
+
+    discovery_summary_json = {
+        "type": 'equal',
+        "title": discovery_title,
+        "value": discovery_equal_value[0]
+    }
+
+    return discovery_summary_json
```
Comment (migraf) [inline pht_federated/aggregator/api/discoveries/utility_functions.py:158]: press str + alt + L in pycharm on all the files for correct formating

File: pht_federated/aggregator/api/endpoints/discovery.py [hunk 27]
```diff
@@ -0,0 +1,117 @@
...
+            "proposal_id": proposal_id,
+            "item_count": discovery_item_count,
+            "feature_count": discovery_feature_count,
+            "column_information": aggregated_feature_lst
+        }
+        discovery_summary = DiscoverySummary(**discovery_summary_schema)
+
+        print("DISCOVERY SUMMARY : {}".format(discovery_summary))
+
+        return discovery_summary
```
Comment (migraf) [inline pht_federated/aggregator/api/endpoints/discovery.py:82]: should be a separate function

File: pht_federated/aggregator/api/endpoints/discovery.py [hunk 28]
```diff
@@ -0,0 +1,117 @@
...
+@router.delete("/{proposal_id}/discovery", response_model=DiscoveryStatistics)
+def delete_discovery_statistics(proposal_id: int, db: Session = Depends(dependencies.get_db)) -> DiscoveryStatistics:
+    discovery_del = datasets.delete_by_proposal_id(proposal_id, db)
+    if not discovery_del:
+        raise HTTPException(status_code=404, detail=f"DatasetStatistics of proposal with id '{proposal_id}' not found.")
+    return discovery_del
+
+@router.post("/{proposal_id}/discovery", response_model=DiscoveryStatistics)
+def post_discovery_statistics(proposal_id: int, create_msg: StatisticsCreate, db: Session = Depends(dependencies.get_db)) -> DatasetStatistics:
+    dataset_statistics = json.loads(create_msg.json())
```
Comment (migraf) [inline pht_federated/aggregator/api/endpoints/discovery.py:94]: avoid unnecessary conversion. Use class methods of pydantic models in this case `create_msg.dict()` or use class fields directly. `json` module should not be used with pydantic models

File: pht_federated/aggregator/api/models/dataset_statistics.py [hunk 29]
```diff
@@ -0,0 +1,23 @@
...
+    feature_count = Column(Integer, default=0)
+    column_information = Column(JSON, default={})
+
+
+class Proposals(Base):
+    __tablename__ = "proposals"
+    id = Column(Integer, primary_key=True, index=True)
+    name = Column(String, unique=True)
+    created_at = Column(DateTime, default=datetime.now())
+    updated_at = Column(DateTime, nullable=True)
```
Comment (migraf) [inline pht_federated/aggregator/api/models/dataset_statistics.py:23]: change id columns to uuid

File: pht_federated/aggregator/db/setup_db.py [hunk 30]
```diff
@@ -0,0 +1,23 @@
+import uuid
+
+from pht_federated.aggregator.storage.db.engine import engine
+from pht_federated.aggregator.db.base import Base
+
+#from pht_federated.aggregator.api.models.discovery import DataSetSummary, DataSet
+from pht_federated.aggregator.api.models import discovery
```
Comment (migraf) [inline pht_federated/aggregator/db/setup_db.py:7]: clean up imports

File: pht_federated/aggregator/storage/db/engine.py [hunk 31]
```diff
@@ -1,5 +1,11 @@
 from sqlmodel import create_engine
+from sqlalchemy.orm import sessionmaker
+
 
 ENGINE_URL = 'postgresql+psycopg2://postgres:postgres@localhost/postgres'
 
+ENGINE_URL = "postgresql+psycopg2://admin:admin@localhost:5442/pht_station"
+
 engine = create_engine(ENGINE_URL)
```
Comment (migraf) [inline pht_federated/aggregator/storage/db/engine.py:9]: make hardcoded connection optional and enable configuration via environment variable

File: pht_federated/tests/aggregator/test_api_discovery.py [hunk 32]
```diff
@@ -0,0 +1,184 @@
...
+        except:
+            print(f"Feature {data['title']} does not have figure_data available.")
+
+    for figure in figure_data_lst:
+        plot_figure_json(figure)
+        #print("Plotting is commented out in 'test_plot_discovery_summary_selected_features'!")
+
+def test_delete_discovery():
+    response = client.delete(f"/api/proposal/{PROPOSAL_ID_NUMERIC}/discovery")
+    assert response.status_code == 200, response.text
```
Comment (migraf) [inline pht_federated/tests/aggregator/test_api_discovery.py:177]: test not only the response code but also whether the results of the aggregation etc. are correct

**Review by migraf — CHANGES_REQUESTED** [review]

**Review by migraf — CHANGES_REQUESTED** [review]

File: pht_federated/aggregator/api/discoveries/utility_functions.py [hunk 33]
```diff
@@ -21,10 +21,7 @@ def aggregate_proposal_features(response: list, proposal_id: int, query: Union[s
     discovery_feature_count = 0
 
     if len(response) < 2:
-        print("Not able to aggregate a discovery summary over less than 2 DatasetStatistics. Aborted.")
-        discovery_summary_schema = {}
-        discovery_summary = DiscoverySummary(**discovery_summary_schema)
-        return discovery_summary
+        raise ValueError("Not able to aggregate a discovery summary over less than 2 DatasetStatistics. Aborted.")
```
Comment (migraf) [inline pht_federated/aggregator/api/discoveries/utility_functions.py:24]: Handle error in endpoint and return appropriate response

**Review by migraf — CHANGES_REQUESTED** [review]

**Review by migraf — APPROVED** [review]
